File: iosApp/iosApp/ConfettiViewModel.swift
import Foundation
import SwiftUI
import ConfettiKit
import AsyncAlgorithms
import KMPNativeCoroutinesAsync


extension SessionDetails: Identifiable { }
extension SpeakerDetails: Identifiable { }
extension RoomDetails: Identifiable { }

enum SessionsUiState {
    case loading
    case success(conferenceName: String, confDates: [Kotlinx_datetimeLocalDate], selectedDateIndex: Int, sessions: [SessionDetails])
}

@MainActor
class ConfettiViewModel: ObservableObject {
    let repository = ConfettiRepository()
    @Published public var speakers: [SpeakerDetails] = []
    @Published public var rooms: [RoomDetails] = []
    
    @Published public var enabledLanguages: Set<String> = []
    
    @Published public var selectedDateIndex: Int = 0
    
    @Published public var uiState: SessionsUiState = .loading
    
    
    init() {
        Task {
            do {
                let stream = asyncStream(for: repository.enabledLanguagesNative)
                for try await data in stream {
                    self.enabledLanguages = data
                }
            } catch {
                print("Failed with error: \(error)")
            }
        }


        Task {
            let conferenceNameAsyncSequence = asyncStream(for: repository.conferenceNameNative)
            let sessionsMapAsyncSequence = asyncStream(for: repository.sessionsMapNative)
            
            for try await (conferenceName, sessionsMap, selectedDateIndex)
                    in combineLatest(conferenceNameAsyncSequence, sessionsMapAsyncSequence, $selectedDateIndex.values) {
                let confDates = sessionsMap.map { $0.key }.sorted { e1, e2 in
                    e2.compareTo(other: e1) > 0
                }
                let selectedDate = confDates[selectedDateIndex]
                let sessions = sessionsMap[selectedDate] ?? []
                self.uiState = SessionsUiState.success(conferenceName: conferenceName, confDates: confDates, selectedDateIndex: selectedDateIndex, sessions: sessions)
            }
        }
    }

    func toggleLanguageChecked(language: String) {
        let checked = enabledLanguages.contains(language) ? false : true
        repository.updateEnableLanguageSetting(language: language, checked: checked)
    }

    
    func observeSpeakers() async {
        do {
            let stream = asyncStream(for: repository.speakersNative)
            for try await data in stream {
                self.speakers = data
            }
        } catch {
            print("Failed with error: \(error)")
        }
    }
    

    
    func observeRooms() async {
        do {
            let stream = asyncStream(for: repository.roomsNative)
            for try await data in stream {
                self.rooms = data
            }
        } catch {
            print("Failed with error: \(error)")
        }
    }
    

    
    func getFlag(session: SessionDetails) -> String {
        // TODO need to figure out how we want to generally handle languages
        return  session.language == "fr-FR" ?  "🇫🇷" : "🇬🇧"
    }
    
    func getSessionSpeakerLocation(session: SessionDetails) -> String {
        var text = ""
        if (session.speakers.count > 0) {
            text = session.speakers.map { $0.speakerDetails.name }.joined(separator: ",")
        }
        text += " (\(session.room?.name ?? ""))"  // \(getFlag(session: session))"
        return text
    }

    
    func getSessionTime(session: SessionDetails) -> String {
        return repository.getSessionTime(session: session)
    }
    
    func refresh() async {
        do {
            try await asyncFunction(for: repository.refreshNative(networkOnly: true))
        } catch {
            print("Failed with error: \(error)")
        }
    }
}

File: androidApp/src/main/java/dev/johnoreilly/confetti/ConfettiViewModel.kt
package dev.johnoreilly.confetti

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dev.johnoreilly.confetti.fragment.SessionDetails
import kotlinx.coroutines.flow.*
import kotlinx.datetime.LocalDate


class ConfettiViewModel(private val repository: ConfettiRepository): ViewModel() {
    val conferenceList = repository.conferenceList
    val speakers = repository.speakers
    val rooms = repository.rooms

    var selectedDateIndex = MutableStateFlow<Int>(0)

    val uiState: StateFlow<SessionsUiState> =
        combine(
            repository.conferenceName,
            repository.sessionsMap,
            selectedDateIndex
        ) { conferenceName, sessionsMap, selectedDateIndex ->
            val confDates = sessionsMap.keys.toList().sorted()
            val selectedDate = confDates[selectedDateIndex]
            val sessions = sessionsMap[selectedDate] ?: emptyList()
            val sessionsByStartTime = sessions.groupBy { repository.getSessionTime(it) }

            SessionsUiState.Success(conferenceName, confDates, selectedDateIndex, sessionsByStartTime)

        }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), SessionsUiState.Loading)


    fun setConference(conference: String) {
        repository.setConference(conference)
    }

    fun switchTab(newIndex: Int) {
        selectedDateIndex.value = newIndex
    }

    fun onLanguageChecked(language: String, checked: Boolean) {
        repository.updateEnableLanguageSetting(language, checked)
    }

    suspend fun getSession(sessionId: String): SessionDetails? {
        return repository.getSession(sessionId)
    }

    suspend fun refresh() {
        repository.refresh()
    }
}

sealed interface SessionsUiState {
    object Loading : SessionsUiState

    data class Success(
        val conferenceName: String,
        val confDates: List<LocalDate>,
        val selectedDateIndex: Int,
        val sessionsByStartTime: Map<String, List<SessionDetails>>
    ) : SessionsUiState
}
